<p align="center">
  <img src="img/GMITLOGO.jpg" width="500"/></p>

# GMIT, H.Dip in Data Analytics, Fundamentals of Data Analysis Project 2018

## 1. Introduction

This repository contains all of the files pertaining to my 2018 project submission for the Fundamentals of Data 
Analysis module of the GMIT H.Dip program in Data Analytics. All of the work within this repository was carried out over the course of a 4 week period in November and December 2018. This Jupyter notebook [1] contains the complete documentation for the project.

### 1.1 Project Objective

The objective of this project is to investigte and explain boxplots and their uses, by researching the subject and then demonstrating the use of box plots using `matplotlib.pyplot`[2]. There are 4 main tasks to carried out [3]:
* Summarise the history of the box plot and situations in which it used.
* Demonstrate the use of the box plot using data of your choosing.
* Explain any relevant terminology such as the terms quartile and percentile. 
* Compare the box plot to alternatives.

### 1.2 Plan for the project

The plan for this project is as follows:
1. Section 2 will give an a brief introduction to boxplots, looking at the history and the use cases. The theory behid them and technical terminology will be explained at this stage
1. In Section 3 I will analyse a data set using boxplots, (insert data set here!)
1. In Section 4 I will discuss some of the findings from Section 3, looking at alternatives to the boxplot, comparing and contrasting the relative strenghts of each approach
1. The project will be concluded in Section 5

## 2. Overview of the Boxplots

### 2.1 Descriptive Statistics

When a statistician or data analyst want to understand and compare datasets, they must have some way way to summarise the dataset. For a one dimensional data set, the data summary should have some idea of where the middle of the dataset lies, and also some measure of what the spread or distribution around this middle value is. A common measure for the middle of a dataset is the *mean*. The mean is calculated by summing the values in the dataset and diving by the number of items in the dataset [4]. The *standard deviation* gives an idea of the spread of the values in the set. The standard deviation is calculated by [5]:
1. Subtracting each value in the data set from the mean
1. Squaring each result
1. Taking the average of these resultant squares
1. Taking the square root of this value

The standard deviation is in the same units as the original data.

A demonstration of how to calculate the mean and standard deviation is shown below:

In [35]:
# Create a list of data
data = [1,2,2,4,5,5,6,7,9,30]

# Calculate mean
total = 0
for i in data:
    total = total + i 
mean = total/len(data)    

# Calculate standard deviation
devsq = 0
for i in data:
    devsq = devsq + (mean - i)**2 
std_dev = (devsq/len(data))**0.5   

print("The mean of the data is : " + str(round(mean,2)))
print("The standard deviation of the data is : " + str(round(std_dev,2)))

The mean of the data is : 7.1
The standard deviation of the data is : 7.98


This demonstrates how to calculate the mean and standard deviation, note however that in this case the mean is 7.1 and only 2 out of the 10 points in the data set are above this value, the other 8 are below it. The mean has been influenced by the extremely high value (30) in the set.

While the mean and standard deviation are useful for describing a dataset - they can be heavily influenced in data has *outliers* [6]. Outliers are extreme values that deviate from other observations in data [7]. If a dataset has outliers, the *five number summary summary* can be used to better describe it.

### 2.2 The five number summary

A different set of descriptive statistics that can be used to summarise a dataset is the *five number summary*. This chooses 5 numbers to help define the centre of the data, and also how well spread out they are [6]. The five numbers are:
1. Minimum - the smallest value in the data set
1. The *first quartile* - also known as the 25th *percentile*, and denoted by *Q<sub>1</sub>*
1. The *median* - this is the midway point of the data set - 50% of the values in the set are below this value
1. The *third quartile* - also know as the 75th *percentile* and denoted by *Q<sub>3</sub>*
1. Maximum - the largest value in the data set

#### Percentiles and Quartiles

The most common definition of the term percentile, is the percentage of values that fall below a certain value [8]. To calculate the *k<sup>th</sup>* percentile (where k is a number between 1 and 100):
1. Order the data set from smallest to largest
1. Multiply *k* percent by the number of points in the data set
1. If the result from step 2 is a whole number go to step 5, otherwise round it to the nearest whole number then go to step 4
1. Count the numbers in you data set from smallest to largest until you get to the one indicated by step 3
1. The corresponding value is the *k<sup>th</sup>* percentile
1. Count the numbers in the data set until you get to the value indicated by step 2, the *k<sup>th</sup>* percentile is the average of that value and the value from the set that follows it

Quartiles are special cases of percentiles:
1. The first quartile, *Q<sub>1</sub>*, is equal to the 25<sup>th</sup> percentile
1. The second quartile (known as the the *median*) is equal to the 50<sup>th</sup> percentile
1. The third quartile, *Q<sub>3</sub>*, is equal to the 75<sup>th</sup> percentile

The the difference between *Q<sub>3</sub>*, and *Q<sub>1</sub>* is known as the *inter-quartile range*, and can be used to give an indication of the spread of the data. The median is used as an estimate for the centre of the data. 

The `pandas` package in Python can be used to calculate the five number summary for the data from the previous example:

In [23]:
# import the package
import pandas as pd
# create a dataframe from the list
df = pd.DataFrame(data, columns=["data"])
# use the describe() function to display the five number summary
df.data.describe()[['min','25%','50%','75%','max']]

min     1.00
25%     2.50
50%     5.00
75%     6.75
max    30.00
Name: data, dtype: float64

The inter-quartile range can be calculated as follows:

In [34]:
# code adapted from https://stackoverflow.com/questions/34782063/how-to-use-pandas-filter-with-iqr

# calculate the interquartile range and print
Q3 = df['data'].quantile(0.75)
Q1 = df['data'].quantile(0.25)

IQR = Q3 - Q1
print("The inter-quartile range for the data is: " + str(IQR))

The inter-quartile range for the data is: 4.25


*Note that the code above uses the **quantile** function - the quantile is another way of saying percentile - except the values are between 0 and 1 - the 25<sup>th</sup> percentile is the 0.25 quantile*

Using the 5 number summary for the data provided - the median is 5, and the interquartile range (IQR) is 4.25. Comparing these to the dataset you can see that 5 is a better estimate for the centre as 50% of the the values are above it and below it. Also an IQR of 4.25 gives a better estimation of the spread as aside from the value of 30 (which look like an outlier) - most of the values fall between plus or minus 4.25 of the median.

### 2.3 The Box Plot

A boxplot is a standardized way of displaying the distribution of data based on a five number summary [10]. It is also a powerful graphical technique for comparing samples from two or more different treatments or populations. It was invented in the 1970s by American statistician John Wilder Tukey [11]. 
It is a compact distributional summary,  displaying less detail than a histogram or kernel density,  but also taking up less space.   Boxplots use robust summary statistics that are always located at actual data points and are quickly computable - even without a computer [12]. Over 40 years later, the box plot has become one of the most widely used statistical plots. In that time the box plot has been developed and enhanced, but Tukey's original box plot would have looked like this:

<p align="center">
  <img src="img/boxplot.png" width="400"/></p>
   <p style="text-align: center;"> <b><I><a href="http://www.physics.csbsju.edu/stats/box2.html">Box Plot</a></I></b> </p>  
   
The boxplot is made up of five components, carefully chosen to give a robust summary of the distribution
of a dataset:
* The median - line is drawn to show this
* The 1<sup>st</sup> and 3<sup>rd</sup> quartiles - depicted as either end of the box
* The maximum and minimum values (know as the extremes)
* Whiskers to join the extremes to the quartiles
* Potential outliers - points that are 1.5 times the interquartile range outside the extremes


### Interpreting a box plot
https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c
https://blog.bioturing.com/2018/05/22/how-to-compare-box-plots/

## 3. How to use the box plot
### Code samples to show how it can be used

## 4. Alternatives
### Comparisons between box plots and other such as histograms

## 6. Conclusion

## References

1. [1] Project Jupyter. Project Jupyter Home. (_https://www.jupyter.org/_)  
1. [2] Matplotlib Development Team. pyplot. (*https://matplotlib.org/api/pyplot_api.html*)
1. [3] Dr. Ian McLoughlin. GMIT. Project 2018: Fundamentals of Data Analysis (_https://github.com/ianmcloughlin/funda-project-2018/raw/master/project.pdf_) 
1. [4] www.mathisfun.com. How to find the mean. (*https://www.mathsisfun.com/mean.html*)
1. [5] www.mathisfun.com. Standard Deviation and Variance. (*https://www.mathsisfun.com/standard-deviation.html*) 
1. [6] www;thoughtco.com What is the five number summary? (https://www.thoughtco.com/what-is-the-five-number-summary-3126237**)
1. [7] www.towardsdatascience.com. A Brief Overview of Outlier Detection Techniques. (https://towardsdatascience.com/a-brief-overview-of-outlier-detection-techniques-1e0b2c19e561)
1. [8] www.statisticshowto.datascience.com. Percentiles, Percentile Rank & Percentile Range (*https://www.statisticshowto.datasciencecentral.com/probability-and-statistics/percentiles-rank-range/*)
1. [9] Deborah J. Rumsey, PhD. Statistics For Dummies 2nd Edition. (pp. 87 - 89). (ISBN - 9781119293521)
1. [10] www.towarddatascience.com. Understanding Boxplots. (*https://towardsdatascience.com/understanding-boxplots-5e2df7bcbd51*)
1. [11] www.britannica.com. Box and Whisker Plot. (*https://www.britannica.com/science/box-and-whisker-plot*)
1. [12] Hadley Wickham and Lisa Stryjewski. 40 years of boxplots. (*http://vita.had.co.nz/papers/boxplots.pdf*)